File: K/k.py
from math import sqrt
def dist(p1, p2):
    d = [abs(p2[0] - p1[0])**2, abs(p2[1] - p1[1])**2, abs(p2[2] - p1[2])**2]
    dist = sqrt(sum(d))
    return dist

if __name__ == "__main__":
    n, k = map(int, input().split())
    planets = []
    for _ in range(n):
        planets.append(list(map(int, input().split())))
    reached = [[0,0,0]]
    while True:
        n = 0
        for r in reached:
            for p in planets:
                if p not in reached and dist(r, p) <= k:
                    reached.append(p)
                    n += 1
        if n == 0:
            break
    print(len(reached) - 1)

File: K/k.cpp
#include <iostream>
#include <initializer_list>
#include <cmath>
#include <vector>
using namespace std;

struct P{
	double x;
	double y;
	double z;
	bool operator==(const P &o) {
		return x == o.x && y == o.y && z == o.z;
	}
};

double dist(P p1, P p2) {
	double dx = (p2.x - p1.x) * (p2.x - p1.x);
	double dy = (p2.y - p1.y) * (p2.y - p1.y);
	double dz = (p2.z - p1.z) * (p2.z - p1.z);
	return sqrt(dx + dy + dz);
}

int main(){ 
	int n, k;
	cin >> n >> k;
	vector<P> planets;
	for (int i = 0; i < n; ++i) {
		double x, y, z;
		cin >> x >> y >> z;
		planets.push_back({x, y, z});
	}
	vector<P> reached {{0, 0, 0}};
	int n_ = 0;
	while (true){ 
		n_  = 0;
		for (P &p : planets) {
			for (P &r : reached) {
				if (p.x == r.x && p.y == r.y && p.z == r.z) {
					break;
				}
				if (dist(p, r) <= k) {
					reached.push_back(p);
					++n_;
					break;
				}
			}
		}
		if (n_ == 0) {
			break;
		}
	}
	cout << reached.size() - 1 << endl;
}
